enhancements/network/on-prem-configurable-lb-dns-stack.md

```diff
@@ -0,0 +1,332 @@
+
+##### Disaster recovery - external LB,DNS
+
+- T0: kube-apiserver recovered
+- T1: kube-apiserver returns OK on /readyz
+- T2: (= T1 + Y Sec , Y=external LB backend health detection time) External LB should mark master-x backend as ready, and should send new connections towards master-x.
+
+### cluster-hosted-net-services operator
+
+The `cluster-hosted-net-services-operator` is a new operator whose operand is a controller for the cluster hosted DNS, LB and VIP services.
+
+### Standard SLO Behavior
+
+The cluster-hosted-net-services-operator is expected to follow the standard behavior of SLO, additionally, since the cluster-hosted-net-services-operator is applicable only to clusters running on on-prem ( 'baremetal', 'openstack', 'vsphere' or 'ovirt') infrastructure,
+It should support also the expected behaviors for "not in use" SLOs similar to the [CBO operator](https://github.com/openshift/enhancements/blob/master/enhancements/baremetal/an-slo-for-baremetal.md#standard-slo-behaviors).
+
+### cluster-hosted-net-services-operator details
+
+The cluster-hosted-net-services-operator should:
+
+- Be a new `openshift/cluster-hosted-net-services-operator` project.
+- Publish an image called `cluster-hosted-net-services-operator`.
+- Add a new `ClusterOperator` with an additional
+  `Disabled` status for non on-prem platforms.
+- Implement a controller reconciling a singleton instance representing the desired cluster hosted LB/DNS/VIP configuration.
+- Do nothing except set `Disabled=true`, `Available=true`, and
+  `Progressing=False` when the `Infrastructure` resource is a platform
+  type other than openstack, baremetal, ovirt or vsphere.
+
+#### CRD definition
+
+Below is a possible option for a CRD instance of this operator:
+
+```yaml
+apiVersion: clusterHostedNetServices.operator.io/v1alpha1
+kind: HostedNetServices
+metadata:
+  name: cluster
+spec:
+  dns: 
+    nodesResolution: Enabled
+    appsResolution: Enabled
+    apiintIpAddress: 192.168.111.5 
+  ingressHa: Enabled
+status:
+  apiServerInternalIpOwner: master-0
+  ingressIpOwner: worker-1
+```
+
+With this sample manifest the cluster-hosted-net-services-operator should:
+
+- Create DNS hosted service that will:
+
+  - resolve node names and .apps wildcard record
+  - resolve api-int to 192.168.111.5.
+  - Run the self-hosted Loadbalancer for api only if apiintIpAddress is equal to API-VIP value provides in install-config file.
+  - Provide high availability for default ingress.
+
+From the status section we can see that master-0 node currently owns api-int VIP and worker-1 node owns ingress VIP. Additional spec and status fields may be introduced based on requirements.
+
+#### Managing  the cluster hosted networking services
+
+All the LB, DNS, and VIP components will be enabled by default in case of fresh deployment as today, while post-deployment the admin can control the services by editing the singleton CR.
```

Comment:
How this will look like in an upgrade scenario? Is there an opt-in for admins to set? If there's no upgrade path how will know that this should not run in such case?

Reply:
If the singleton instance exists --> it's upgrade path else fresh deployment.